### PR TITLE
`copilot-c99`: Use `InitArray` for nested arrays and structs. Refs #276.

### PR DESCRIPTION
**Pull request by RyanGlScott**
Previously, C99 translation would produce initializers for array and struct values using `InitExpr`. When arrays or structs are nested inside of other arrays or structs, this causes the generated code to be produce warnings when compiled with `gcc -Wpedantic` due to the placement of the casts in the initializer expressions. In the worst case, these arrays could even cause C compilers to perform unwanted type conversions. For the full story, see the comments in the `constinit` function.

To fix the issue, we now include special cases for `Array` and `Struct` such that they are translated using `InitArray`. For example, this means that a nested array will be translated as `{{0, 1}, {2, 3}, {4, 5}}` instead of `{ (int32_t[2]){0, 1}, (int32_t[2]){2, 3}, (int32_t[2]){4, 5} }`. The same syntax also works for structs—for example, an array of structs will be translated as `{ {0, 1, 2} }` instead of `{ (struct foo){0, 1, 2} }`. The use of `InitArray` is a bit odd (one might imagine `{ { .a = 0, .b = 1, .c = 2 } }` instead), but that is an artifact of how `language-c99-simple`'s API works.

Bottom line: fewer casts involving array or struct types, which gives C compilers fewer headaches.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
I've verified that with this patch, the generated code for both of the examples in https://github.com/Copilot-Language/copilot/issues/276#issuecomment-1100158705 compiles without warnings with `gcc -Wpedantic` and `clang -Wpedantic`. I also tried a nested structs example:

```hs
{-# LANGUAGE DataKinds        #-}
{-# LANGUAGE RebindableSyntax #-}
module Main where

import Copilot.Compile.C99
import Language.Copilot

data SampleStruct2 = SampleStruct2
  { sampleStruct2Num :: Field "sample_struct2_num" Int32
  }

instance Struct SampleStruct2 where
  typename _ = "sample_struct2_t"
  toValues sampleStruct2 = [ Value typeOf (sampleStruct2Num sampleStruct2) ]

instance Typed SampleStruct2 where
  typeOf = Struct (SampleStruct2 (Field 1))

data SampleStruct = SampleStruct
  { sampleStructField :: Field "sample_struct_field" SampleStruct2
  }

instance Struct SampleStruct where
  typename _ = "sample_struct_t"
  toValues sampleStruct = [ Value typeOf (sampleStructField sampleStruct) ]

instance Typed SampleStruct where
  typeOf = Struct (SampleStruct (Field (SampleStruct2 (Field 1))))

main :: IO ()
main = do
  spec' <- reify spec
  compile "nestedstructs" spec'

spec :: Spec
spec = do
  trigger "handlerA" true [arg arr]
  trigger "handlerS2" true [arg v2]
  trigger "handlerS" true [arg v1]

arr :: Stream Int32
arr = [ 1 ] ++ arr

v2 :: Stream SampleStruct2
v2 = [ SampleStruct2 (Field 1) ] ++ v2

v1 :: Stream SampleStruct
v1 = [ SampleStruct (Field (SampleStruct2 (Field 1))) ] ++ v1
```
```c
// Generated code
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "nestedstructs.h"

struct sample_struct2_t
{ int32_t sample_struct2_num;
};
struct sample_struct_t
{ struct sample_struct2_t sample_struct_field;
};
static int32_t s0[(1)] = {((int32_t)(1))};
static struct sample_struct2_t s1[(1)] = {{((int32_t)(1))}};
static struct sample_struct_t s2[(1)] = {{{((int32_t)(1))}}};

// <snip>
```

As well as a nested arrays example:

```hs
{-# LANGUAGE RebindableSyntax #-}
{-# LANGUAGE DataKinds        #-}

module Main where

import Copilot.Compile.C99
import Language.Copilot

arr :: Stream (Array 2 (Array 2 Int32))
arr = [ array [array [0, 1], array [2, 3]] ] ++ arr

spec :: Spec
spec = do
  trigger "func" true [arg arr]

main :: IO ()
main = do
  spec' <- reify spec
  compile "nestedarrays" spec'
```
```c
// Generated code
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "nestedarrays.h"

static int32_t s0[(1)][(2)][(2)] = {{{((int32_t)(0)), ((int32_t)(1))}, {((int32_t)(2)), ((int32_t)(3))}}};

// <snip>
```
